### PR TITLE
#993: Popover should handle context  (closes #993)

### LINKED ISSUE #993: Popover should handle context
## requirements
When `attachToBody` is passed to `Popover` it creates a new render tree without passing the received context. 

## specs
- add a prop to `Popover` to describe the context
- wrap `Popover`'s children in a component responsible for assigning the context

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by ascariandrea**
Closes #993

### tests performed
- tested on alinity-pro ✅ 

- [ ] Google Chrome
- [ ] Internet Explorer

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/popover/Popover.js [hunk 1]
```diff
@@ -69,7 +72,8 @@ export default class Popover extends React.Component {
   componentDidUpdate() {
     if (this.containerNode) {
       const popover = this.getVisiblePopover();
-      ReactDOM.render(popover, this.containerNode);
+      const ContextWrapper = getContextWrapper(this.props.popover.contextTypes);
```
Comment (FrancescoCioria) [inline src/popover/Popover.js]: you should use `this.getPopoverProps()` instead of `this.props`

File: src/popover/Popover.js [hunk 2]
```diff
@@ -69,7 +72,8 @@ export default class Popover extends React.Component {
...
+      ReactDOM.render(<ContextWrapper context={this.props.popover.context}>{popover}</ContextWrapper>, this.containerNode);
```
Comment (FrancescoCioria) [inline src/popover/Popover.js]: here `context` is optional but in `ContextWrapper` is required

File: src/popover/Popover.js [hunk 3]
```diff
@@ -236,7 +240,8 @@ export default class Popover extends React.Component {
 
     // render invisible popover
     const hiddenPopover = this.getHiddenPopover();
-    ReactDOM.render(hiddenPopover, this.containerNode);
+    const ContextWrapper = getContextWrapper(this.props.popover.contextTypes);
+    ReactDOM.render(<ContextWrapper context={this.props.popover.context}>{hiddenPopover}</ContextWrapper>, this.containerNode);
```
Comment (FrancescoCioria) [inline src/popover/Popover.js]: same

File: src/popover/examples/hover-auto-attach-to-body.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+const Option = ({ title, value }) => (
```
Comment (FrancescoCioria) [inline src/popover/examples/hover-auto-attach-to-body.js]: this file shouldn't be here 😛

File: src/utils/index.js [hunk 5]
```diff
@@ -1,5 +1,27 @@
+import React from 'react';
 import cx from 'classnames';
+import { props, t } from 'tcomb-react';
 
 export skinnable, { contains } from 'react-skinnable';
-export { props, t } from 'tcomb-react';
+export { props, t };
 export const stateClassUtil = (...classes) => cx([].concat(...classes).map(cl => `is-${cl}`));
+
+export const getContextWrapper = (contextTypes) => {
```
Comment (giogonzo) [inline src/utils/index.js]: minor, but the same exact thing I think is already implemented for Modal here: https://github.com/buildo/react-components/blob/master/src/modal/modalWithContext.js
What about reusing/sharing?

Comment (giogonzo) [inline src/utils/index.js]: no, probably not exactly the same.. if not straightforward, 👍 as it is

File: src/popover/examples/hover-auto-attach-to-body.js (hunk 4 above)
Comment (ascariandrea) [inline src/popover/examples/hover-auto-attach-to-body.js]: Removed 🙈

File: src/utils/index.js (hunk 5 above)
Comment (ascariandrea) [inline src/utils/index.js]: 👍

**Pull request merged**